# Combo Sku Analysis

In [2]:
import pandas as pd

In [3]:
# Analyze combo_sku.csv
print("📦 COMBO SKU ANALYSIS")
print("="*50)

try:
    combo_df = pd.read_csv('combo_sku.csv')
    print(f"✅ combo_sku.csv loaded successfully!")
    print(f"📊 Shape: {combo_df.shape}")
    print(f"📋 Columns: {list(combo_df.columns)}")
    
    # Show first few rows
    print(f"\n📋 First 5 rows:")
    print(combo_df.head())
    
    # Basic info
    print(f"\n🔍 Data Info:")
    print(combo_df.info())
    
    # Check for missing values
    print(f"\n❓ Missing Values:")
    missing_combo = combo_df.isnull().sum()
    print(missing_combo[missing_combo > 0])
    
except FileNotFoundError:
    print("❌ combo_sku.csv not found!")
    print("📁 Please ensure the file is in the current directory")
    print("🔍 Expected: Information about product combinations/bundles")
    
except Exception as e:
    print(f"❌ Error loading combo_sku.csv: {e}")



📦 COMBO SKU ANALYSIS
✅ combo_sku.csv loaded successfully!
📊 Shape: (910, 17)
📋 Columns: ['Combo ', 'SKU1', 'SKU2', 'SKU3', 'SKU4', 'SKU5', 'SKU6', 'SKU7', 'SKU8', 'SKU9 ', 'SKU10', 'SKU11', 'SKU12', 'SKU13', 'SKU14', 'Status', 'Unnamed: 16']

📋 First 5 rows:
                                Combo                               SKU1  \
0   CSTE_0341_Wands_Hermoine_Sunglasas        CSTE_0291_OT_Hermoine_wand   
1    CSTE_0359_Wands_Hermoine_light_Sg    CSTE_0331_Wands_Hermoine_Light   
2    CSTE_0354_Wands_Hermoine_light_MB    CSTE_0331_Wands_Hermoine_Light   
3    CSTE_0335_Wands_Hermoine_MusicBox        CSTE_0291_OT_Hermoine_wand   
4  CSTE_0358_Wands_Dumbledore_light_Sg  CSTE_0333_Wands_Dumbledore_Light   

                             SKU2 SKU3 SKU4 SKU5 SKU6 SKU7 SKU8  SKU9   SKU10  \
0  CSTE_0319_SG_HarryPotter_Round  NaN  NaN  NaN  NaN  NaN  NaN    NaN    NaN   
1  CSTE_0319_SG_HarryPotter_Round  NaN  NaN  NaN  NaN  NaN  NaN    NaN    NaN   
2           CSTE_0020_MB_HP_Black  NaN  N

In [4]:
import pandas as pd
import numpy as np

# Load the sku_mappings.csv file
df_sku_mappings = pd.read_csv('sku_mappings.csv')

# Display basic info about the file
print("Shape of sku_mappings.csv:", df_sku_mappings.shape)
print("\nColumn names:")
print(df_sku_mappings.columns.tolist())
print("\nFirst 5 rows:")
print(df_sku_mappings.head())

Shape of sku_mappings.csv: (5218, 12)

Column names:
['sku', 'msku', 'panels', 'Status', 'Status.1', 'Unnamed: 5', 'image', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 11']

First 5 rows:
        sku      msku         panels    Status      Status.1  Unnamed: 5  \
0  15694321  15694321  Rudrav Meesho  Inactive         block         NaN   
1  23654985  23654985  Rudrav Meesho  Inactive  Combo,Paused         NaN   
2  28547595  28547595  Rudrav Meesho  Inactive        paused         NaN   
3  30258741  30258741  Rudrav Meesho       NaN           NaN         NaN   
4  30548764  30548764  Rudrav Meesho  Inactive         block         NaN   

  image Unnamed: 7  Unnamed: 8 Unnamed: 9 Unnamed: 10 Unnamed: 11  
0   NaN        NaN         NaN        NaN         NaN         NaN  
1   NaN        NaN         NaN        NaN         NaN         NaN  
2   NaN        NaN         NaN        NaN         NaN         NaN  
3   NaN        NaN         NaN        NaN         NaN       

In [5]:
# Check for missing values in each column
print("Missing values per column:")
print(df_sku_mappings.isnull().sum())
print("\n" + "="*50)

# Check data types
print("Data types:")
print(df_sku_mappings.dtypes)
print("\n" + "="*50)

# Look at unique values in key columns
print("Unique values in 'panels' column:")
print(df_sku_mappings['panels'].value_counts())
print("\n" + "="*50)

print("Unique values in 'Status' column:")
print(df_sku_mappings['Status'].value_counts(dropna=False))

Missing values per column:
sku               0
msku            548
panels            0
Status          337
Status.1       2032
Unnamed: 5     5218
image          2271
Unnamed: 7     1707
Unnamed: 8     5218
Unnamed: 9     5211
Unnamed: 10    5198
Unnamed: 11    5204
dtype: int64

Data types:
sku             object
msku            object
panels          object
Status          object
Status.1        object
Unnamed: 5     float64
image           object
Unnamed: 7      object
Unnamed: 8     float64
Unnamed: 9      object
Unnamed: 10     object
Unnamed: 11     object
dtype: object

Unique values in 'panels' column:
panels
CSTE AMAZON      1535
CSTE MEESHO      1360
Rudrav Meesho     834
CSTE FK           801
GL FK             687
Cste Amazon         1
Name: count, dtype: int64

Unique values in 'Status' column:
Status
Active         3298
Inactive        951
In Progress     632
NaN             337
Name: count, dtype: int64


In [6]:
# Remove columns that are mostly empty (Unnamed columns)
columns_to_keep = ['sku', 'msku', 'panels', 'Status', 'Status.1', 'image']
df_cleaned = df_sku_mappings[columns_to_keep].copy()

print("After removing unnecessary columns:")
print("Shape:", df_cleaned.shape)
print("\nMissing values:")
print(df_cleaned.isnull().sum())
print("\n" + "="*50)

# Look at rows with missing msku values
print("Rows with missing MSKU (first 10):")
missing_msku = df_cleaned[df_cleaned['msku'].isnull()]
print(missing_msku.head(10))
print(f"\nTotal rows with missing MSKU: {len(missing_msku)}")

After removing unnecessary columns:
Shape: (5218, 6)

Missing values:
sku            0
msku         548
panels         0
Status       337
Status.1    2032
image       2271
dtype: int64

Rows with missing MSKU (first 10):
           sku msku       panels Status Status.1 image
4668   2054025  NaN  CSTE MEESHO    NaN      NaN   NaN
4669  Cj_0Izuv  NaN  CSTE MEESHO    NaN      NaN   NaN
4670   CSTE071  NaN  CSTE MEESHO    NaN      NaN   NaN
4671   CSTE088  NaN  CSTE MEESHO    NaN      NaN   NaN
4672   CSTE131  NaN  CSTE MEESHO    NaN      NaN   NaN
4673   CSTE141  NaN  CSTE MEESHO    NaN      NaN   NaN
4674   CSTE168  NaN  CSTE MEESHO    NaN      NaN   NaN
4675   CSTE169  NaN  CSTE MEESHO    NaN      NaN   NaN
4676   CSTE177  NaN  CSTE MEESHO    NaN      NaN   NaN
4677   CSTE191  NaN  CSTE MEESHO    NaN      NaN   NaN

Total rows with missing MSKU: 548


In [7]:
# Look at the missing MSKU rows grouped by panel
print("Missing MSKU breakdown by panel:")
print(missing_msku['panels'].value_counts())
print("\n" + "="*50)

# Look at more examples of missing MSKUs to see if there's a pattern
print("More examples of missing MSKUs:")
print(missing_msku[['sku', 'panels']].head(20))
print("\n" + "="*50)

# Check if any SKUs that start with "CSTE" have MSKUs in other rows
cste_skus_with_msku = df_cleaned[df_cleaned['sku'].str.startswith('CSTE', na=False) & df_cleaned['msku'].notna()]
print("Examples of CSTE SKUs that DO have MSKUs:")
print(cste_skus_with_msku[['sku', 'msku']].head(10))

Missing MSKU breakdown by panel:
panels
CSTE MEESHO      377
CSTE AMAZON       70
Rudrav Meesho     60
CSTE FK           40
GL FK              1
Name: count, dtype: int64

More examples of missing MSKUs:
           sku       panels
4668   2054025  CSTE MEESHO
4669  Cj_0Izuv  CSTE MEESHO
4670   CSTE071  CSTE MEESHO
4671   CSTE088  CSTE MEESHO
4672   CSTE131  CSTE MEESHO
4673   CSTE141  CSTE MEESHO
4674   CSTE168  CSTE MEESHO
4675   CSTE169  CSTE MEESHO
4676   CSTE177  CSTE MEESHO
4677   CSTE191  CSTE MEESHO
4678   CSTE194  CSTE MEESHO
4679   CSTE196  CSTE MEESHO
4680   CSTE202  CSTE MEESHO
4681   CSTE203  CSTE MEESHO
4682   CSTE210  CSTE MEESHO
4683   CSTE211  CSTE MEESHO
4684   CSTE222  CSTE MEESHO
4685   CSTE223  CSTE MEESHO
4686   CSTE225  CSTE MEESHO
4687   CSTE227  CSTE MEESHO

Examples of CSTE SKUs that DO have MSKUs:
                                    sku                                msku
251       CSTE-0007-ST-BTS-PILLOW-COOKY       CSTE_0007_ST_Bts_Pillow_Cooky
265       CST

In [8]:
# Remove rows with missing MSKUs since they can't be used for mapping
df_cleaned = df_cleaned[df_cleaned['msku'].notna()].copy()

print("After removing missing MSKU rows:")
print("Shape:", df_cleaned.shape)
print("Missing values now:")
print(df_cleaned.isnull().sum())
print("\n" + "="*50)

# Check for duplicate SKU mappings (same SKU mapped to different MSKUs)
print("Checking for duplicate SKUs with different MSKUs:")
sku_counts = df_cleaned['sku'].value_counts()
duplicate_skus = sku_counts[sku_counts > 1]
print(f"Number of SKUs that appear multiple times: {len(duplicate_skus)}")

if len(duplicate_skus) > 0:
    print("\nFirst 10 duplicate SKUs:")
    for sku in duplicate_skus.head(10).index:
        print(f"\nSKU: {sku}")
        print(df_cleaned[df_cleaned['sku'] == sku][['sku', 'msku', 'panels']])

After removing missing MSKU rows:
Shape: (4670, 6)
Missing values now:
sku            0
msku           0
panels         0
Status        50
Status.1    1726
image       1854
dtype: int64

Checking for duplicate SKUs with different MSKUs:
Number of SKUs that appear multiple times: 440

First 10 duplicate SKUs:

SKU: CSTE_0381_ST_Rabbit_Doll_Grey
                                sku                           msku  \
2463  CSTE_0381_ST_Rabbit_Doll_Grey  CSTE_0381_ST_Rabbit_Doll_Grey   
2464  CSTE_0381_ST_Rabbit_Doll_Grey  CSTE_0381_ST_Rabbit_Doll_Grey   
2466  CSTE_0381_ST_Rabbit_Doll_Grey  CSTE_0381_ST_Rabbit_Doll_Grey   
2467  CSTE_0381_ST_Rabbit_Doll_Grey  CSTE_0381_ST_Rabbit_Doll_Grey   
2468  CSTE_0381_ST_Rabbit_Doll_Grey  CSTE_0381_ST_Rabbit_Doll_Grey   

             panels  
2463    CSTE MEESHO  
2464  Rudrav Meesho  
2466    CSTE AMAZON  
2467          GL FK  
2468        CSTE FK  

SKU: CSTE_0382_ST_Rabbit_Doll_Pink
                                sku                           msk

In [9]:
# Check if any SKU maps to DIFFERENT MSKUs (this would be a problem)
problematic_mappings = []

for sku in duplicate_skus.index:
    sku_data = df_cleaned[df_cleaned['sku'] == sku]
    unique_mskus = sku_data['msku'].unique()
    
    if len(unique_mskus) > 1:
        problematic_mappings.append(sku)

print(f"SKUs that map to DIFFERENT MSKUs (problematic): {len(problematic_mappings)}")

if len(problematic_mappings) > 0:
    print("Problematic SKUs:")
    for sku in problematic_mappings[:5]:  # Show first 5
        print(f"\nSKU: {sku}")
        print(df_cleaned[df_cleaned['sku'] == sku][['sku', 'msku', 'panels']])
else:
    print("✅ Good! All duplicate SKUs map to the same MSKU across different panels")
    
print("\n" + "="*50)
print("Summary of current data:")
print(f"Total rows: {len(df_cleaned)}")
print(f"Unique SKUs: {df_cleaned['sku'].nunique()}")
print(f"Unique MSKUs: {df_cleaned['msku'].nunique()}")
print("\nPanel distribution:")
print(df_cleaned['panels'].value_counts())

SKUs that map to DIFFERENT MSKUs (problematic): 19
Problematic SKUs:

SKU: greycirclewatch
                  sku                               msku         panels
3954  greycirclewatch    CSTE_O0318_OT_circlewatch_Black  Rudrav Meesho
3958  greycirclewatch    CSTE_O0318_OT_circlewatch_Black    CSTE MEESHO
3970  greycirclewatch  CSTE_O0321_OT_Square watches_Grey          GL FK
3971  greycirclewatch  CSTE_O0321_OT_Square watches_Grey        CSTE FK

SKU: CSTE_0407_OT_HP_Scarf_SG
                           sku                             msku         panels
2510  CSTE_0407_OT_HP_Scarf_SG         CSTE_0407_OT_HP_Scarf_SG    CSTE MEESHO
2511  CSTE_0407_OT_HP_Scarf_SG         CSTE_0407_OT_HP_Scarf_SG  Rudrav Meesho
4093  CSTE_0407_OT_HP_Scarf_SG  CSTE_O0362_OT_HarryPotter_Scarf          GL FK
4094  CSTE_0407_OT_HP_Scarf_SG  CSTE_O0362_OT_HarryPotter_Scarf        CSTE FK

SKU: CSTE_0147_SG_ButterFly_Green
                               sku                             msku  \
1409  CSTE_0147_S

In [10]:
# Show all 19 problematic SKUs with their conflicting mappings
print("All 19 problematic SKUs with conflicting MSKU mappings:")
print("="*70)

for i, sku in enumerate(problematic_mappings, 1):
    print(f"\n{i}. SKU: {sku}")
    problem_data = df_cleaned[df_cleaned['sku'] == sku][['sku', 'msku', 'panels']]
    print(problem_data.to_string(index=False))
    print("-" * 50)

# Count how many rows these problematic mappings affect
problematic_rows = df_cleaned[df_cleaned['sku'].isin(problematic_mappings)]
print(f"\nTotal rows affected by problematic mappings: {len(problematic_rows)}")
print(f"This represents {len(problematic_rows)/len(df_cleaned)*100:.1f}% of your data")

All 19 problematic SKUs with conflicting MSKU mappings:

1. SKU: greycirclewatch
            sku                              msku        panels
greycirclewatch   CSTE_O0318_OT_circlewatch_Black Rudrav Meesho
greycirclewatch   CSTE_O0318_OT_circlewatch_Black   CSTE MEESHO
greycirclewatch CSTE_O0321_OT_Square watches_Grey         GL FK
greycirclewatch CSTE_O0321_OT_Square watches_Grey       CSTE FK
--------------------------------------------------

2. SKU: CSTE_0407_OT_HP_Scarf_SG
                     sku                            msku        panels
CSTE_0407_OT_HP_Scarf_SG        CSTE_0407_OT_HP_Scarf_SG   CSTE MEESHO
CSTE_0407_OT_HP_Scarf_SG        CSTE_0407_OT_HP_Scarf_SG Rudrav Meesho
CSTE_0407_OT_HP_Scarf_SG CSTE_O0362_OT_HarryPotter_Scarf         GL FK
CSTE_0407_OT_HP_Scarf_SG CSTE_O0362_OT_HarryPotter_Scarf       CSTE FK
--------------------------------------------------

3. SKU: CSTE_0147_SG_ButterFly_Green
                         sku                            msku        pa

In [11]:
# Go back to the original cleaned data (before removing "problematic" mappings)
df_cleaned_correct = df_cleaned.copy()

# Create a composite key of SKU + Panel
df_cleaned_correct['sku_panel_key'] = df_cleaned_correct['sku'].astype(str) + "_" + df_cleaned_correct['panels'].astype(str)

print("Checking for TRUE duplicates (same SKU + Panel combination):")
print(f"Total rows: {len(df_cleaned_correct)}")
print(f"Unique SKU+Panel combinations: {df_cleaned_correct['sku_panel_key'].nunique()}")

# Find actual duplicates (same SKU on same panel appearing multiple times)
sku_panel_counts = df_cleaned_correct['sku_panel_key'].value_counts()
true_duplicates = sku_panel_counts[sku_panel_counts > 1]

print(f"True duplicates (same SKU + Panel appearing multiple times): {len(true_duplicates)}")

if len(true_duplicates) > 0:
    print("\nFirst 5 true duplicates:")
    for sku_panel in true_duplicates.head().index:
        print(f"\nSKU+Panel: {sku_panel}")
        duplicate_rows = df_cleaned_correct[df_cleaned_correct['sku_panel_key'] == sku_panel]
        print(duplicate_rows[['sku', 'msku', 'panels', 'Status']].to_string(index=False))
        print("-" * 50)

Checking for TRUE duplicates (same SKU + Panel combination):
Total rows: 4670
Unique SKU+Panel combinations: 4659
True duplicates (same SKU + Panel appearing multiple times): 8

First 5 true duplicates:

SKU+Panel: UnicornSlipper_WhiteSmall_Rudrav Meesho
                      sku                                    msku        panels Status
UnicornSlipper_WhiteSmall CSTE_O0006_OT_UnicornSlipper_WhiteSmall Rudrav Meesho Active
UnicornSlipper_WhiteSmall CSTE_O0006_OT_UnicornSlipper_WhiteSmall Rudrav Meesho Active
UnicornSlipper_WhiteSmall CSTE_O0006_OT_UnicornSlipper_WhiteSmall Rudrav Meesho Active
UnicornSlipper_WhiteSmall CSTE_O0006_OT_UnicornSlipper_WhiteSmall Rudrav Meesho Active
--------------------------------------------------

SKU+Panel: PILLOW_PACK2_GL FK
         sku                        msku panels Status
PILLOW_PACK2 CSTE_O0043_OT_CushionFiller  GL FK Active
PILLOW_PACK2 CSTE_O0043_OT_CushionFiller  GL FK Active
PILLOW_PACK2 CSTE_O0043_OT_CushionFiller  GL FK Active
--------

In [12]:
# Remove duplicate rows (keep only the first occurrence of each SKU+Panel combination)
df_final_clean = df_cleaned_correct.drop_duplicates(subset=['sku', 'panels'], keep='first')

print("After removing true duplicates:")
print(f"Original rows: {len(df_cleaned_correct)}")
print(f"Clean rows: {len(df_final_clean)}")
print(f"Duplicate rows removed: {len(df_cleaned_correct) - len(df_final_clean)}")

print(f"\nUnique SKUs: {df_final_clean['sku'].nunique()}")
print(f"Unique MSKUs: {df_final_clean['msku'].nunique()}")
print(f"Unique SKU+Panel combinations: {df_final_clean.groupby(['sku', 'panels']).size().shape[0]}")

# Final verification - should be no more duplicates
final_sku_panel_counts = df_final_clean.groupby(['sku', 'panels']).size()
remaining_duplicates = final_sku_panel_counts[final_sku_panel_counts > 1]
print(f"\nRemaining duplicates after cleaning: {len(remaining_duplicates)}")

print("\nFinal panel distribution:")
print(df_final_clean['panels'].value_counts())

# Drop the temporary composite key column
df_final_clean = df_final_clean.drop('sku_panel_key', axis=1)
print("\n✅ Dataset is now clean and ready!")

After removing true duplicates:
Original rows: 4670
Clean rows: 4659
Duplicate rows removed: 11

Unique SKUs: 3912
Unique MSKUs: 1162
Unique SKU+Panel combinations: 4659

Remaining duplicates after cleaning: 0

Final panel distribution:
panels
CSTE AMAZON      1465
CSTE MEESHO       982
Rudrav Meesho     771
CSTE FK           761
GL FK             679
Cste Amazon         1
Name: count, dtype: int64

✅ Dataset is now clean and ready!


In [13]:
# Save the clean dataset
df_final_clean.to_csv('sku_msku_mapping_clean.csv', index=False)
print("✅ Clean dataset saved as 'sku_msku_mapping_clean.csv'")

print("\n" + "="*60)
print("FINAL CLEAN DATASET SUMMARY")
print("="*60)
print(f"📊 Total records: {len(df_final_clean):,}")
print(f"🏪 Marketplaces: {df_final_clean['panels'].nunique()}")
print(f"📦 Unique SKUs: {df_final_clean['sku'].nunique():,}")
print(f"🏷️  Unique MSKUs: {df_final_clean['msku'].nunique():,}")
print(f"🔗 Unique SKU+Panel combinations: {len(df_final_clean):,}")

print(f"\n📈 Marketplace Distribution:")
panel_dist = df_final_clean['panels'].value_counts()
for panel, count in panel_dist.items():
    percentage = (count/len(df_final_clean))*100
    print(f"   {panel}: {count:,} ({percentage:.1f}%)")

print(f"\n🧹 Data Quality:")
print(f"   ✅ No duplicate SKU+Panel combinations")
print(f"   ✅ No missing SKUs or MSKUs")
print(f"   ✅ Clean 1:1 mapping within each marketplace")

print(f"\n💡 Key Insights:")
print(f"   • Same SKU can map to different MSKUs across marketplaces")
print(f"   • SKU + Panel combination is the unique identifier")
print(f"   • Ready for cost analysis and calculations")

print("\nSample of clean data:")
print(df_final_clean.head()[['sku', 'msku', 'panels']].to_string(index=False))

✅ Clean dataset saved as 'sku_msku_mapping_clean.csv'

FINAL CLEAN DATASET SUMMARY
📊 Total records: 4,659
🏪 Marketplaces: 6
📦 Unique SKUs: 3,912
🏷️  Unique MSKUs: 1,162
🔗 Unique SKU+Panel combinations: 4,659

📈 Marketplace Distribution:
   CSTE AMAZON: 1,465 (31.4%)
   CSTE MEESHO: 982 (21.1%)
   Rudrav Meesho: 771 (16.5%)
   CSTE FK: 761 (16.3%)
   GL FK: 679 (14.6%)
   Cste Amazon: 1 (0.0%)

🧹 Data Quality:
   ✅ No duplicate SKU+Panel combinations
   ✅ No missing SKUs or MSKUs
   ✅ Clean 1:1 mapping within each marketplace

💡 Key Insights:
   • Same SKU can map to different MSKUs across marketplaces
   • SKU + Panel combination is the unique identifier
   • Ready for cost analysis and calculations

Sample of clean data:
     sku     msku        panels
15694321 15694321 Rudrav Meesho
23654985 23654985 Rudrav Meesho
28547595 28547595 Rudrav Meesho
30258741 30258741 Rudrav Meesho
30548764 30548764 Rudrav Meesho


## Here's how I cleaned the sku_mappings.csv file:

1. Initial Data Inspection: Loaded the CSV and examined its structure, finding 6 columns with 5,218 total rows. Identified that the key columns for mapping were 'sku', 'msku', and 'panels' (marketplaces).
2. Handle Missing Values: Found 548 rows with missing MSKU values, primarily from "CSTE MEESHO" marketplace entries. Removed these rows since they couldn't be used for SKU-to-MSKU mapping, leaving 4,670 usable records.
3. Understand Data Model: Initially thought duplicate SKUs were problematic, but realized that 'panels' represent different marketplaces. Understood that the same SKU can legitimately map to different MSKUs across different marketplaces.
4. Identify True Duplicates: Created a composite key using SKU+Panel combination to find real duplicates. Found only 8 true duplicates where the same SKU appeared multiple times on the same marketplace.
5. Final Cleaning: Removed the 11 duplicate rows (keeping first occurrence of each SKU+Panel combination). Saved the final clean dataset with 4,659 rows, ensuring each SKU+Panel combination appears exactly once.



